File: swift/Sources/DSA/DataStructures/Stack.swift
/// LIFO (Last In, First Out) data structure.
/// Time: O(1) for all operations
/// Space: O(n)
public struct Stack<T> {
    private var items: [T] = []

    public init() {}

    public mutating func push(_ value: T) {
        items.append(value)
    }

    public mutating func pop() -> T? {
        items.popLast()
    }

    public func peek() -> T? {
        items.last
    }

    public var isEmpty: Bool {
        items.isEmpty
    }

    public var count: Int {
        items.count
    }

    public mutating func clear() {
        items.removeAll()
    }

    public func toArray() -> [T] {
        items
    }
}

File: java/src/main/java/com/dsa/datastructures/Stack.java
package com.dsa.datastructures;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * A generic stack implementation using ArrayList.
 * Time: O(1) for push, pop, peek
 * Space: O(n)
 */
public class Stack<T> {
    private final List<T> data;

    public Stack() {
        this.data = new ArrayList<>();
    }

    public void push(T value) {
        data.add(value);
    }

    public Optional<T> pop() {
        if (data.isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(data.remove(data.size() - 1));
    }

    public Optional<T> peek() {
        if (data.isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(data.get(data.size() - 1));
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    public int size() {
        return data.size();
    }

    public void clear() {
        data.clear();
    }
}
